File: scripts/timeGrabber.ts
import { JSDOM } from "jsdom";
import { penalties } from "@/data";
import {
  type SeasonName,
  type TrackName,
  drivers,
  type RacerName,
} from "@/types";

const formatDate = (dateString: string) => {
  const date = new Date(dateString);
  const monthNames = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
  ];
  const day = date.getUTCDate();
  const month = monthNames[date.getUTCMonth()];
  const year = date.getUTCFullYear();

  const getOrdinal = (n: number) => {
    const s = ["th", "st", "nd", "rd"];
    const v = n % 100;
    return s[(v - 20) % 10] || s[v] || s[0];
  };

  return `${month} ${day}${getOrdinal(day)}, ${year}`;
};

export const timeGrabber = async (
  season: SeasonName,
  race: TrackName,
  link: string
) => {
  const response = await fetch(link);
  const html = await response.text();
  const dom = new JSDOM(
    html
      .replace(/<style([\S\s]*?)>([\S\s]*?)<\/style>/gim, "")
      .replace(/<script([\S\s]*?)>([\S\s]*?)<\/script>/gim, "")
  );

  const dateElement =
    dom.window.document.querySelector("h1")?.nextElementSibling?.textContent;
  const date = dateElement ? formatDate(dateElement) : null;

  const table = dom.window.document.querySelector("table");

  const results: Record<string, string> = {};
  const fastestLap = {
    racerId: "unknown",
    time: "23:59:999",
  };
  let highestLapCount = 0;

  if (table) {
    const times = Array.from(table.querySelectorAll("tr"))
      .map((row) => {
        const cells = row.querySelectorAll("td");
        const driver = cells[1];
        const totalTime = cells[3];
        const numLaps = cells[4];
        const bestLap = cells[5];

        if (driver && totalTime && bestLap && numLaps) {
          return [
            driver.innerHTML.split("\n")[0],
            totalTime.innerHTML,
            bestLap.innerHTML.split("<")[0],
            numLaps.innerHTML.split("lap")[0],
          ];
        }

        return [];
      })
      .slice(1);

    let lapCounter = 0;

    times.forEach((time, index) => {
      const tableName = time[0];
      const name =
        Object.entries(drivers).find((driver) =>
          (driver[1] as { tableNames?: string[] }).tableNames?.includes(
            tableName
          )
        )?.[0] ?? tableName;

      let parsedTime = time[1].split(":").slice(1).join(":");

      if (Number(time[3]) > highestLapCount) {
        highestLapCount = Number(time[3]);
      } else if (Number(time[3]) <= highestLapCount - 5) {
        parsedTime = "DNF";
      } else {
        if (index > 0) {
          parsedTime = (
            (Date.parse(`01/01/2000 ${time[1]}`) -
              Date.parse(`01/01/2000 ${times[index - 1][1]}`)) /
            1000
          )
            .toFixed(3)
            .toString();
        }

        if (parsedTime.startsWith("-")) {
          lapCounter++;
          parsedTime = `${lapCounter} lap${lapCounter === 1 ? "" : "s"}`;
        } else if (lapCounter > 0) {
          parsedTime = `${lapCounter} lap${lapCounter === 1 ? "" : "s"}`;
        }
      }

      results[name] = parsedTime;

      if (
        time[2] !== "00:00.000" &&
        Date.parse(`01/01/2000 00:${time[2]}`) <
          Date.parse(`01/01/2000 00:${fastestLap.time}`)
      ) {
        fastestLap.time = time[2];
        fastestLap.racerId = name;
      }
    });
  }

  // Perform any penalty rearrangements
  const resultsKeys = Object.keys(results) as RacerName[];
  const copy = [...resultsKeys];
  copy.forEach((result, index) => {
    const penalty = penalties[season]?.[race]?.[result];

    if (penalty) {
      resultsKeys.splice(index, 1);
      resultsKeys.splice(index + penalty, 0, result);
    }
  });

  const orderedResults = resultsKeys.reduce(
    (obj: Record<string, string>, item: string) => {
      return (obj[item] = results[item]), obj;
    },
    {}
  );

  return { results: orderedResults, fastestLap, data: link, date };
};

File: scripts/time_grabber.rs
use std::collections::HashMap;
use reqwest;
use scraper::{Html, Selector};
use chrono::{DateTime, Utc};
use serde_json::json;

fn format_date(date_string: &str) -> String {
    let date = DateTime::parse_from_rfc3339(date_string).unwrap();
    let month_names = [
        "January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December",
    ];
    let day = date.day();
    let month = month_names[date.month0() as usize];
    let year = date.year();

    let get_ordinal = |n: u32| -> &str {
        let s = ["th", "st", "nd", "rd"];
        let v = n % 100;
        s[(v - 20) % 10] || s[v] || s[0]
    };

    format!("{} {}{}, {}", month, day, get_ordinal(day), year)
}

pub async fn time_grabber(season: &str, race: &str, link: &str) -> serde_json::Value {
    let response = reqwest::get(link).await.unwrap().text().await.unwrap();
    let document = Html::parse_document(&response);

    let date_selector = Selector::parse("h1 + *").unwrap();
    let date_element = document.select(&date_selector).next().map(|e| e.inner_html());
    let date = date_element.map(|d| format_date(&d));

    let table_selector = Selector::parse("table").unwrap();
    let table = document.select(&table_selector).next();

    let mut results: HashMap<String, String> = HashMap::new();
    let mut fastest_lap = json!({
        "racerId": "unknown",
        "time": "23:59:999"
    });
    let mut highest_lap_count = 0;

    if let Some(table) = table {
        let rows = table.select(&Selector::parse("tr").unwrap()).skip(1);
        let mut lap_counter = 0;

        for (index, row) in rows.enumerate() {
            let cells: Vec<_> = row.select(&Selector::parse("td").unwrap()).collect();
            let driver = cells.get(1).map(|e| e.inner_html()).unwrap_or_default();
            let total_time = cells.get(3).map(|e| e.inner_html()).unwrap_or_default();
            let num_laps = cells.get(4).map(|e| e.inner_html()).unwrap_or_default();
            let best_lap = cells.get(5).map(|e| e.inner_html()).unwrap_or_default();

            let name = drivers.iter().find(|(_, driver)| driver.table_names.contains(&driver)).map(|(name, _)| name).unwrap_or(&driver).to_string();

            let mut parsed_time = total_time.split(':').skip(1).collect::<Vec<_>>().join(":");

            if num_laps.parse::<i32>().unwrap_or(0) > highest_lap_count {
                highest_lap_count = num_laps.parse::<i32>().unwrap_or(0);
            } else if num_laps.parse::<i32>().unwrap_or(0) <= highest_lap_count - 5 {
                parsed_time = "DNF".to_string();
            } else {
                if index > 0 {
                    let previous_time = rows.get(index - 1).map(|r| r.select(&Selector::parse("td").unwrap()).nth(3).map(|e| e.inner_html()).unwrap_or_default()).unwrap_or_default();
                    parsed_time = ((DateTime::parse_from_rfc3339(&format!("2000-01-01T{}Z", total_time)).unwrap().timestamp_millis() - DateTime::parse_from_rfc3339(&format!("2000-01-01T{}Z", previous_time)).unwrap().timestamp_millis()) / 1000).to_string();
                }

                if parsed_time.starts_with('-') {
                    lap_counter += 1;
                    parsed_time = format!("{} lap{}", lap_counter, if lap_counter == 1 { "" } else { "s" });
                } else if lap_counter > 0 {
                    parsed_time = format!("{} lap{}", lap_counter, if lap_counter == 1 { "" } else { "s" });
                }
            }

            results.insert(name.clone(), parsed_time);

            if best_lap != "00:00.000" && DateTime::parse_from_rfc3339(&format!("2000-01-01T00:{}Z", best_lap)).unwrap().timestamp_millis() < DateTime::parse_from_rfc3339(&format!("2000-01-01T00:{}Z", fastest_lap["time"].as_str().unwrap())).unwrap().timestamp_millis() {
                fastest_lap["time"] = json!(best_lap);
                fastest_lap["racerId"] = json!(name);
            }
        }
    }

    let results_keys: Vec<_> = results.keys().cloned().collect();
    let mut copy = results_keys.clone();
    for (index, result) in copy.iter().enumerate() {
        if let Some(penalty) = grid_penalties.get(season).and_then(|r| r.get(race)).and_then(|p| p.get(result)) {
            results_keys.remove(index);
            results_keys.insert((index as i32 + penalty) as usize, result.clone());
        }
    }

    let ordered_results: HashMap<_, _> = results_keys.into_iter().map(|k| (k.clone(), results[&k].clone())).collect();

    json!({
        "results": ordered_results,
        "fastestLap": fastest_lap,
        "data": link,
        "date": date
    })
}
